### PR TITLE
Feat/c api updates

### PR DESCRIPTION
**Pull request by mchekhovoi**
OTA-3212 SendDeviceData()/SendManifest()/OpenStoredTarget() C APIs

### CODE REVIEW

**Comment by OYTIS** [comment]
This pull request **introduces 1 alert** when merging ba59e760cab0ac51f91944921b3fedeacd63f050 into 74fa9fdaefe39412d05e0c0ad1f96d64af78b66b - [view on LGTM.com](https://lgtm.com/projects/g/advancedtelematic/aktualizr/rev/pr-6b4e02541f8bc9c548aba02ab3dac95bd480aa76)

**new alerts:**

* 1 for Return c\_str of local std::string

---

**Warning** - Automated code review for advancedtelematic/aktualizr will be disabled on October 1, 2019. You can avoid this by <a href="https://github.com/apps/lgtm-com/installations/new">installing the LGTM.com GitHub App</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/libaktualizr-c/libaktualizr-c.cc [hunk 1]
```diff
@@ -135,3 +137,70 @@ int Aktualizr_install_target(Aktualizr *a, Target *t) {
...
+  try {
+    a->SendDeviceData();
+    return 0;
+  } catch (const std::exception &e) {
+    std::cerr << "Aktualizr_send_device_data exception: " << e.what() << std::endl;
+    return -1;
+  }
+}
+
+void *Aktualizr_open_stored_target(Aktualizr *a, const char *filename, const char *content) {
```
Comment (lbonn) [inline src/libaktualizr-c/libaktualizr-c.cc]: The C api already defines a `Target` type, can we use it here as a parameter instead of re-parsing Json? Also the `content` parameter name is quite non-descriptive.

File: src/libaktualizr-c/libaktualizr-c.cc [hunk 2]
```diff
@@ -135,3 +137,70 @@ int Aktualizr_install_target(Aktualizr *a, Target *t) {
...
+  try {
+    auto handle = a->OpenStoredTarget(target);
+    return reinterpret_cast<void *>(handle.release());
+  } catch (const std::exception &e) {
+    std::cerr << "Aktualizr_open_stored_target exception: " << e.what() << std::endl;
+    return nullptr;
+  }
+}
+
+size_t Aktualizr_read_stored_target(void *handle, uint8_t *buf, size_t size) {
```
Comment (lbonn) [inline src/libaktualizr-c/libaktualizr-c.cc]: Even if it does not give type safety, we could define a dummy proxy type for handle instead of `void*`, it would make the api more self-documenting.

File: src/libaktualizr-c/test/api-test.c [hunk 3]
```diff
@@ -27,7 +27,7 @@ int main(int argc, char **argv) {
 
   c = Aktualizr_campaigns_check(a);
   if (c) {
-    printf("Accepting running campaign\n");
```
Comment (lbonn) [inline src/libaktualizr-c/test/api-test.c]: Why the change from `printf` to `fprintf(stdout, ...)`?

File: src/libaktualizr-c/libaktualizr-c.cc (hunk 1 above)
Comment (mchekhovoi) [reply src/libaktualizr-c/libaktualizr-c.cc]: Fixed.

File: src/libaktualizr-c/libaktualizr-c.cc (hunk 2 above)
Comment (mchekhovoi) [reply src/libaktualizr-c/libaktualizr-c.cc]: Fixed.

File: src/libaktualizr-c/test/api-test.c (hunk 3 above)
Comment (mchekhovoi) [reply src/libaktualizr-c/test/api-test.c]: Reverted.

File: src/libaktualizr-c/test/CMakeLists.txt [hunk 4]
```diff
@@ -5,6 +5,8 @@ SET(CMAKE_SKIP_RPATH TRUE)
 
 add_executable(${TARGET_NAME} EXCLUDE_FROM_ALL ${SOURCES})
 add_dependencies(build_tests ${TARGET_NAME})
-target_link_libraries(${TARGET_NAME} aktualizr-c)
+target_link_libraries(${TARGET_NAME} aktualizr-c api-test-utils)
 
 aktualizr_source_file_checks(${SOURCES})
+
+add_subdirectory(api-test-utils)
```
Comment (lbonn) [inline src/libaktualizr-c/test/CMakeLists.txt]: To remove these tests from the default target:

```suggestion
add_subdirectory(api-test-utils EXCLUDE_FROM_ALL)
```

File: src/libaktualizr-c/test/api-test-utils/api-test-utils.cc [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  TestUtils::waitForServer(server + "/");
+}
+
+UptaneGenerator *Get_uptane_generator(const char *path) { return new Process(path); }
+
+void Run_uptane_generator(UptaneGenerator *generator, const char **args, size_t args_count) {
+  std::vector<std::string> args_vector;
+  args_vector.reserve(args_count);
+  for (size_t i = 0; i < args_count; ++i) {
+    args_vector.emplace_back(args[i]);
```
Comment (lbonn) [inline src/libaktualizr-c/test/api-test-utils/api-test-utils.cc]: It's getting an hard-to-avoid lint warning:

```suggestion
    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
    args_vector.emplace_back(args[i]);
```

File: src/libaktualizr-c/test/api-test-utils/api-test-utils.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
+#ifndef API_TEST_UTILS_H
+#define API_TEST_UTILS_H
+
+#ifdef __cplusplus
+#include "../../aktualizr/src/libaktualizr/config/config.h"
```
Comment (lbonn) [inline src/libaktualizr-c/test/api-test-utils/api-test-utils.h]: We try to avoid these error-prone relative includes. config.h is directly accessible by `#include "config/config.h"`. For test_utils.h, it should be solveable by playing with CMake include directories for some test targets.

We can maybe define a CMake function in the spirit of `add_aktualizr_test`, dedicated for C tests, if needed. It can probably wait until we start having two or three of these, though.

File: src/libaktualizr-c/test/CMakeLists.txt (hunk 4 above)
Comment (mchekhovoi) [reply src/libaktualizr-c/test/CMakeLists.txt]: Fixed.

File: src/libaktualizr-c/test/api-test-utils/api-test-utils.cc (hunk 5 above)
Comment (mchekhovoi) [reply src/libaktualizr-c/test/api-test-utils/api-test-utils.cc]: Outdated.
Anyway thanks for the explanation.

File: src/libaktualizr-c/test/api-test-utils/api-test-utils.h (hunk 6 above)
Comment (mchekhovoi) [reply src/libaktualizr-c/test/api-test-utils/api-test-utils.h]: Fixed by CMake include directories.
add_aktualizr_test function is not added for now.

File: src/libaktualizr-c/test/api-test-utils/api-test-utils.cc [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  auto *server_handle = new boost::process::child(serverPath, port, "-f", "-m", metaPath);
+  TestUtils::waitForServer(serverAddress + "/");
+
+  return server_handle;
+}
+
+void Stop_fake_http_server(FakeHttpServer *server) {
+  // NOLINTNEXTLINE
+  if (server != nullptr && server->valid()) {
+    // NOLINTNEXTLINE
```
Comment (patrickvacek) [inline src/libaktualizr-c/test/api-test-utils/api-test-utils.cc]: Can you add the linter ID here so that we know exactly what is being exempted?

File: src/libaktualizr/campaign/campaign_test.cc [hunk 8]
```diff
@@ -31,6 +31,47 @@ TEST(campaign, Campaigns_from_json) {
   EXPECT_FALSE(campaignsNoAutoAccept.at(0).autoAccept);
 }
 
+/* Get JSON from campaign. */
+TEST(campaign, Campaign_to_json) {
+  auto json = Utils::parseJSONFile(test_data_dir / "campaigns_sample.json");
+
+  auto campaigns = campaign::campaignsFromJson(json);
```
Comment (lbonn) [inline src/libaktualizr/campaign/campaign_test.cc]: These two tests seem redundant. Furthermore, `Campaign_to_json` actually uses `campaign::campaignsFromJson`. We can probably just keep one of the two.

Comment (mchekhovoi) [reply src/libaktualizr/campaign/campaign_test.cc]: Fixed. Only "Campaigns_to_json" test is remained.
